### PR TITLE
Update fuzzy_match.py

### PR DESCRIPTION
**Pull request by ivyleavedtoadflax**
# Description

Updates the `FuzzyMatch` class by removing the reliance on pandas, instead using a indices and lists of dicts. This approach cuts the time taken to match 620 references against a list of 5 million publication from a ~10-15s search time to ~3-5s search time, reducing time taken from over two hours to just under one.

## Type of change

Please delete options that are not relevant.

- [x] :sparkles: New feature (refactor)
- [x] :memo: Documentation update

# How Has This Been Tested?

`make # docker tests`

Also, manually labelled titles extracted from the validation for the 2019.10.8 model were matched against the 5 million publications.

# Checklist:

- [x] My code follows the style guidelines of this project (pep8 AND pyflakes)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I included tests in my PR
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: reach/refparse/utils/fuzzy_match.py [hunk 1]
```diff
@@ -1,53 +1,108 @@
...
+                "Matched title": best_match.get("title"),
+                "Matched publication id": _str_or_null(best_match.get("uber_id")),
+                "Matched publication pmcid": _str_or_null(best_match.get("pmcid")),
+                "Matched publication pmid": _str_or_null(best_match.get("pmid")),
+                "Matched publication doi": _str_or_null(best_match.get("doi")),
+                "Similarity": best_similarity,
+                "Match algorithm": "Fuzzy match",
             }
+
+def _str_or_null(docid):
```
Comment (nsorros) [inline reach/refparse/utils/fuzzy_match.py]: why is this needed?

File: reach/refparse/utils/fuzzy_match.py [hunk 2]
```diff
@@ -1,53 +1,108 @@
...
         if best_similarity > self.similarity_threshold:
             return {
-                'Document id': reference['Document id'],
-                'Reference id': reference['Reference id'],
-                'Extracted title': reference['Title'],
-                'Matched title': best_match['title'],
-                'Matched publication id': best_match['uber_id'],
-                'Similarity': best_similarity,
-                'Match algorithm': 'Fuzzy match'
+                "Document id": _str_or_null(reference.get("Document id")),
```
Comment (nsorros) [inline reach/refparse/utils/fuzzy_match.py]: a good question to ask is why would all these information be missing if there is a match? and is that a behaviour we want?

File: reach/refparse/utils/fuzzy_match.py [hunk 3]
```diff
@@ -1,53 +1,108 @@
...
+            return None
+
+        if len(reference.get("Title")) < self.title_length_threshold:
+            return None
 
         retrieved_publications = self.search_publications(reference)
 
         best_match = retrieved_publications.iloc[0]
-        best_similarity = best_match['similarity']
+        best_similarity = best_match.get("similarity")
```
Comment (nsorros) [inline reach/refparse/utils/fuzzy_match.py]: isn't there a problem if there is no similarity?

File: reach/refparse/utils/fuzzy_match.py [hunk 4]
```diff
@@ -1,53 +1,108 @@
...
+
+        Args:
+            publications(list): List of dicts containing publication info as
+                above, or a pandas.DataFrame.
+            similarity_threshold(float): Minimum allowable similarity
+                threshold. If not results are found to have a similarity
+                higher than this, then nothing is returned.
+            title_length_threhold(float): Minimum allowed length of title.
+                Less than this threshold will return no results.
+        """
```
Comment (nsorros) [inline reach/refparse/utils/fuzzy_match.py]: super helpful, thanks for adding

File: reach/refparse/utils/fuzzy_match.py [hunk 5]
```diff
@@ -1,53 +1,108 @@
...
-        if len(reference['Title']) < self.title_length_threshold:
-            return
+            return None
+
+        if len(reference.get("Title")) < self.title_length_threshold:
+            return None
 
         retrieved_publications = self.search_publications(reference)
 
         best_match = retrieved_publications.iloc[0]
```
Comment (lizgzil) [inline reach/refparse/utils/fuzzy_match.py]: FYI if you haven't seen - there is an issue open about this: https://github.com/wellcometrust/reach/issues/165 Please contribute to the discussion if you have any opinions about it

File: reach/refparse/utils/fuzzy_match.py [hunk 6]
```diff
@@ -1,53 +1,108 @@
...
+                "Matched publication id": _str_or_null(best_match.get("uber_id")),
+                "Matched publication pmcid": _str_or_null(best_match.get("pmcid")),
+                "Matched publication pmid": _str_or_null(best_match.get("pmid")),
+                "Matched publication doi": _str_or_null(best_match.get("doi")),
+                "Similarity": best_similarity,
+                "Match algorithm": "Fuzzy match",
             }
+
+def _str_or_null(docid):
+    if docid:
```
Comment (lizgzil) [inline reach/refparse/utils/fuzzy_match.py]: So if docid = None, then `_str_or_null` will return the string "None"? Just wondering whether this will mess up anything down the line, might be worth keeping it as None. Or is there a reason to convert it to a string?

File: reach/refparse/utils/fuzzy_match.py (hunk 1 above)
Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: I was having some issues where the pmid (which looks like an int) was being converted to string as: `4576358.0`. When the data is coerced to `pandas.DataFrame` the column has type `float` applied to it. I could fix it in pandas, but because it wasn't entirely clear to me whether the input `publications` should be a dataframe or a list of dicts, I added a function here.

File: reach/refparse/utils/fuzzy_match.py (hunk 2 above)
Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: Good question. I probably need to re-add the document id to the manually tagged data.

File: reach/refparse/utils/fuzzy_match.py (hunk 3 above)
Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: Yep you're right, this is probably not required

File: reach/refparse/utils/fuzzy_match.py [hunk 7]
```diff
@@ -1,53 +1,108 @@
...
+        ]
+        ```
+
+        Args:
+            publications(list): List of dicts containing publication info as
+                above, or a pandas.DataFrame.
+            similarity_threshold(float): Minimum allowable similarity
+                threshold. If not results are found to have a similarity
+                higher than this, then nothing is returned.
+            title_length_threhold(float): Minimum allowed length of title.
```
[the remark this answers is not in the document]
Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: yep, good spot!

File: reach/refparse/utils/fuzzy_match.py (hunk 6 above)
Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: The intended function is: if the `id` is a `float`, then convert to `int` before returning as `str`.

This is a good question. This partly comes back to the `list` of `dict`s/`pd.DataFrame` question - what should the input of `publications` be. If the latter, then we can probably just sanitize the column types in the DataFrame itself.

File: reach/refparse/utils/fuzzy_match.py (hunk 4 above)
Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: Also on this @nsorros, I have set the threshold as a `float` (0:1), but in various places I have seen it as an integer. Which is the correct?

File: reach/refparse/tests/test_fuzzy_match.py [hunk 8]
```diff
@@ -1,38 +1,39 @@
...
+from reach.refparse.utils import FuzzyMatcher
 
 
 class TestFuzzyMatcherInit(unittest.TestCase):
     def test_empty_title(self):
         real_publications = pd.DataFrame({
             'title': []
         })
-        threshold = 75
+        threshold = 0.75
```
Comment (ivyleavedtoadflax) [inline reach/refparse/tests/test_fuzzy_match.py]: Is this right  @lizgzil @nsorros? Should these be decimals or integers?

File: reach/refparse/utils/fuzzy_match.py (hunk 4 above)
Comment (nsorros) [reply reach/refparse/utils/fuzzy_match.py]: that's a good question, let me have a look

File: reach/refparse/tests/test_fuzzy_match.py (hunk 8 above)
Comment (nsorros) [reply reach/refparse/tests/test_fuzzy_match.py]: i think that is right, it seems like a "bug"

Comment (nsorros) [reply reach/refparse/tests/test_fuzzy_match.py]: https://github.com/wellcometrust/reach/blob/df709550f9bdcb7653cc0912e13047d9768f6604/reach/refparse/settings.py#L10

File: reach/refparse/utils/fuzzy_match.py (hunk 4 above)
Comment (nsorros) [reply reach/refparse/utils/fuzzy_match.py]: https://github.com/wellcometrust/reach/blob/df709550f9bdcb7653cc0912e13047d9768f6604/reach/refparse/settings.py#L10

File: reach/refparse/utils/fuzzy_match.py (hunk 2 above)
Comment (nsorros) [reply reach/refparse/utils/fuzzy_match.py]: This is the policy id so without it you will not be able to match, or you could but you would need to map the titles back to the ids.

File: reach/refparse/utils/fuzzy_match.py [hunk 9]
```diff
@@ -37,38 +37,71 @@ def __init__(
...
                 higher than this, then nothing is returned.
-            title_length_threhold(float): Minimum allowed length of title.
+            title_length_threhold(int): Minimum allowed length of title.
                 Less than this threshold will return no results.
         """
-        # Filter out any publications that don't have titles
+        # Filter out any publications that don't have titles assuming that the
+        # input is a jsonl.
 
-        self.publications = pd.DataFrame(publications)
```
Comment (nsorros) [inline reach/refparse/utils/fuzzy_match.py]: so these changes are mainly to eliminate the need for pandas?

File: reach/refparse/utils/fuzzy_match.py [hunk 10]
```diff
@@ -37,38 +37,71 @@ def __init__(
...
+            doc.update({"similarity": similarity_vector[key]})
+
+        # Convert to a list, again checking that sorting is correct
+
+        retrieved_publications = [value for key, value in retrieved_publications.items()]
+
+
+        retrieved_publications = sorted(retrieved_publications, key=lambda x: x['similarity'], reverse=True)
+
+        return retrieved_publications
```
Comment (nsorros) [inline reach/refparse/utils/fuzzy_match.py]: i am not 100% convinced that this is a better implementation. it substitutes 6 lines for 27. what is the main motivation behind it?

Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: The old implementation was making a deepcopy of the publications dataframe every time it did a match. This dataframe has 5 million rows, so this step considerably slows the process. On my machine it locks up >20GB of RAM and maxes out one core (it's not parralelisable) for about 11-15 seconds per query. On the current task I of matching manual labels, that is ~620 * ~13 seconds = 2.23 hours. This new implementation takes ~620 * ~4 seconds = 0.688 hours. Speed is the only motivation!

File: reach/refparse/utils/fuzzy_match.py (hunk 9 above)
Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: See the following comment. These changes are for speed, and the existing implementation was inefficient because it created deepcopies of a very large dataframe, and it also executed repeated searches across the whole dataframe. By relying on indexes, I've managed to eliminate the slow search.

**Comment by ivyleavedtoadflax** [comment]
Implemeted in https://github.com/wellcometrust/reach/pull/314/commits/53d98ac59f032ab51164ba0a3678a99cac070f73.

**Comment by ivyleavedtoadflax** [comment]
> I think we need to have some fields as required not everything is optional for fuzzy matching to make sense

Implemeted in 53d98ac.

File: reach/refparse/utils/fuzzy_match.py [hunk 11]
```diff
@@ -1,53 +1,136 @@
...
+
+        # Cycle through just the similar docs, not the entire list of
+        # publications.
+
+        retrieved_publications = {i:self.publications[i] for i in similar_doc_indices}
+
+        # Finally append the similarity to the similar docs based on its index
+
+        for key, doc in retrieved_publications.items():
+            doc.update({"similarity": similarity_vector[key]})
```
Comment (nsorros) [inline reach/refparse/utils/fuzzy_match.py]: I am not sure why you need the intermediate key i everywhere. can you not just zip publications and similarities?

```
for similarity, publication in zip(sorted_similarities, publications)
    publication['similarity'] = similarity
```

File: reach/refparse/utils/fuzzy_match.py (hunk 3 above)
Comment (nsorros) [reply reach/refparse/utils/fuzzy_match.py]: Again, why would that be a get? we need a similarity, otherwise fuzzy matching does not make sense.

File: reach/refparse/utils/fuzzy_match.py [hunk 12]
```diff
@@ -1,53 +1,136 @@
...
-                'Extracted title': reference['Title'],
-                'Matched title': best_match['title'],
-                'Matched publication id': best_match['uber_id'],
-                'Similarity': best_similarity,
-                'Match algorithm': 'Fuzzy match'
+                "Document id": reference["Document id"],
+                "Reference id": reference["Reference id"],
+                "Extracted title": reference["Title"],
+                "Matched title": best_match["title"],
+                "Matched publication id": best_match.get("uber_id"),
```
Comment (nsorros) [inline reach/refparse/utils/fuzzy_match.py]: i think you need one of those ids, otherwise what did you match with?

Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: True, but we don't know which ones are going to be missing

File: reach/refparse/utils/fuzzy_match.py (hunk 11 above)
Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: We could just zip them together, but that will mean that we need to traverse the entire publications list, and that is what I am trying to avoid. This way we can more efficiently extract the indices of interest here: `retrieved_publications = {i:self.publications[i] for i in similar_doc_indices}` - we only traverse the `similar_doc_indices`, not the entire `self.publications`.

File: reach/refparse/utils/fuzzy_match.py (hunk 3 above)
Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: yep, sorry meant to change this. It could be a []

File: reach/refparse/utils/fuzzy_match.py (hunk 12 above)
Comment (nsorros) [reply reach/refparse/utils/fuzzy_match.py]: one of them needs to be there, it is the unique identifier of the publication whatever that is. what i am trying to say is that the tool assume you pass in a list of publications with a unique identifier so you need to maintain that functionality

Comment (ivyleavedtoadflax) [reply reach/refparse/utils/fuzzy_match.py]: At present one or more can be absent from the publications data, so I would rather get all three, accept that some night be missing, and then work on creating a better if (maybe a hash of all three IDs) in the next step.

Comment (nsorros) [reply reach/refparse/utils/fuzzy_match.py]: Which publication data are you referring to?

File: reach/refparse/utils/fuzzy_match.py [hunk 13]
```diff
@@ -1,53 +1,136 @@
-import numpy as np
 import logging
 
+import numpy as np
+
```
Comment (nsorros) [inline reach/refparse/utils/fuzzy_match.py]: remove newline

File: reach/refparse/utils/fuzzy_match.py (hunk 10 above)
Comment (nsorros) [reply reach/refparse/utils/fuzzy_match.py]: 👍

File: reach/refparse/utils/fuzzy_match.py (hunk 3 above)
Comment (nsorros) [reply reach/refparse/utils/fuzzy_match.py]: cool, good to go when you do the change
